File: Assets/CloudCode/SubmitTournamentResult.js
// SubmitTournamentResult.js
// Deploy to UGS Cloud Code (same module as the other Arena scripts).
//
// PURPOSE — closes the "client writes its own leaderboard score" hole.
// The client no longer calls Leaderboards.AddPlayerScore directly. Instead it
// submits the FULL final standings it computed (the ordered list of entryIds,
// champion first) plus its own claimed placement. The server:
//
//   1. verifies the caller actually belongs to the tournament (via the
//      server-authored tournament_player_map_{weekId}) and that the claimed
//      entryId / placement are internally consistent with the submitted
//      standings;
//   2. cross-checks the submitted standings against OTHER real players in the
//      same bracket. Because every honest client resolves the identical
//      deterministic bracket (same snapshots, same seeds), honest submissions
//      produce an identical standings hash. A lone cheater's forged standings
//      will not match the plurality, so it is rejected;
//   3. once a standings hash reaches quorum (or the bracket has too few real
//      players for quorum to ever exist), the server itself writes the weekly
//      placement score and maintains an authoritative all-time championship
//      counter — the client never writes leaderboards.
//
// Deploy alongside a dashboard Access-Control policy that DENIES player tokens
// write access to the arena_weekly / arena_alltime leaderboards, so this
// endpoint (running under the service context) is the only writer.
//
// Game Data entities used (all project-scoped shared custom data):
//   "tournament_player_map_{weekId}" — key: playerId  → { tournamentId, entryId, scoreBand }   (written by LockAndAssignBrackets)
//   "tournament_brackets_{weekId}"   — key: tournamentId → bracket blob (has realPlayerCount)   (written by LockAndAssignBrackets)
//   "tournament_results_{weekId}"    — key: playerId  → { tournamentId, standingsHash, placement, entryId, submittedUtc }
//   "tournament_scored_{weekId}"     — key: playerId  → "done"   (idempotency guard: score already written)
//   "arena_championships"            — key: playerId  → integer career championship count (authoritative)
//
// Parameters:
//   weekId        (STRING)  — client-computed week ID (e.g., "W20260413")
//   tournamentId  (STRING)  — the bracket the player claims to have finished
//   entryId       (STRING)  — the player's own entryId within that bracket
//   placement     (NUMERIC) — claimed final placement (1 = champion)
//   standingsJson (STRING)  — JSON array of entryIds ordered by placement (index 0 = 1st)
//
// Returns:
//   { status: "scored",  placement, score }        — verified and written
//   { status: "pending", have, need }              — stored, waiting for corroboration; client should retry later
//   { status: "already_scored" }                    — idempotent no-op
//   { status: "rejected", reason }                   — failed validation / consensus

const { DataApi } = require("@unity-services/cloud-save-1.4");

// ── Leaderboard IDs (must match ArenaLeaderboardService.cs) ──
const WEEKLY_LEADERBOARD_ID = "arena_weekly";
const ALLTIME_LEADERBOARD_ID = "arena_alltime";

const BRACKET_SIZE = 32;

// Minimum number of real players (including the submitter) whose independently
// computed standings must AGREE before a result is trusted. Brackets whose real
// player count is below this can never reach quorum, so they fall back to
// single-submission trust (flagged in logs) — a lone real player beating 31 bots.
const CONSENSUS_QUORUM = 2;

module.exports = async ({ params, context, logger }) => {
  const { projectId, playerId } = context;
  const { weekId, tournamentId, entryId, placement, standingsJson } = params;

  // ── Parameter validation ──
  if (!weekId || typeof weekId !== "string") {
    return { status: "rejected", reason: "Invalid week ID." };
  }
  if (!tournamentId || typeof tournamentId !== "string") {
    return { status: "rejected", reason: "Invalid tournament ID." };
  }
  if (!entryId || typeof entryId !== "string") {
    return { status: "rejected", reason: "Invalid entry ID." };
  }
  if (
    typeof placement !== "number" ||
    placement < 1 ||
    placement > BRACKET_SIZE
  ) {
    return { status: "rejected", reason: "Invalid placement." };
  }
  if (!standingsJson || typeof standingsJson !== "string") {
    return { status: "rejected", reason: "Standings are required." };
  }

  let standings;
  try {
    standings = JSON.parse(standingsJson);
  } catch (e) {
    return { status: "rejected", reason: "Malformed standings." };
  }
  if (!Array.isArray(standings) || standings.length === 0) {
    return { status: "rejected", reason: "Standings must be a non-empty array." };
  }

  const cloudSave = new DataApi(context);

  // ── Idempotency: has this player already been scored for the week? ──
  const scoredEntity = `tournament_scored_${weekId}`;
  if (await hasItem(cloudSave, projectId, scoredEntity, playerId, logger)) {
    return { status: "already_scored" };
  }

  // ── Authority check: the player must be mapped to THIS tournament by the
  //    server-authored bracket assignment, with THIS entryId. ──
  const mapEntity = `tournament_player_map_${weekId}`;
  const mapping = await getItem(cloudSave, projectId, mapEntity, playerId, logger);
  if (!mapping) {
    return { status: "rejected", reason: "You are not registered for this week." };
  }
  let mapObj;
  try {
    mapObj = JSON.parse(mapping);
  } catch (e) {
    return { status: "rejected", reason: "Server mapping unreadable." };
  }
  if (mapObj.tournamentId !== tournamentId || mapObj.entryId !== entryId) {
    return { status: "rejected", reason: "Tournament/entry mismatch." };
  }

  // ── Internal consistency: the claimed placement must match the position of
  //    the player's own entryId inside the submitted standings. ──
  const claimedIndex = standings.indexOf(entryId);
  if (claimedIndex === -1) {
    return { status: "rejected", reason: "Your entry is absent from the standings." };
  }
  if (claimedIndex + 1 !== placement) {
    return {
      status: "rejected",
      reason: "Claimed placement disagrees with submitted standings.",
    };
  }

  // ── Store this submission (keyed by playerId in the shared results entity) ──
  const standingsHash = hashString(JSON.stringify(standings));
  const resultsEntity = `tournament_results_${weekId}`;
  try {
    await cloudSave.setPrivateCustomItem(projectId, resultsEntity, {
      key: playerId,
      value: JSON.stringify({
        tournamentId,
        standingsHash,
        placement,
        entryId,
        submittedUtc: Math.floor(Date.now() / 1000),
      }),
    });
  } catch (e) {
    logger.error(`Failed to store result for ${playerId}: ${e.message}`);
    return { status: "rejected", reason: "Server error saving result." };
  }

  // ── Gather all submissions for this tournament and tally standings hashes ──
  const allResults = await getAllItems(cloudSave, projectId, resultsEntity, logger);
  let matchingVotes = 0;
  const hashVotes = {};
  for (const r of allResults) {
    let obj;
    try {
      obj = JSON.parse(r.value);
    } catch (e) {
      continue;
    }
    if (obj.tournamentId !== tournamentId) continue;
    hashVotes[obj.standingsHash] = (hashVotes[obj.standingsHash] || 0) + 1;
    if (obj.standingsHash === standingsHash) matchingVotes++;
  }

  // Plurality hash across all real submissions for this bracket.
  let bestHash = null;
  let bestVotes = 0;
  for (const h of Object.keys(hashVotes)) {
    if (hashVotes[h] > bestVotes) {
      bestVotes = hashVotes[h];
      bestHash = h;
    }
  }

  // How many real players are in this bracket? Small brackets can never reach
  // quorum, so we allow single-submission trust for them (but log it).
  const realPlayerCount = await getBracketRealPlayerCount(
    cloudSave,
    projectId,
    weekId,
    tournamentId,
    logger,
  );
  const effectiveQuorum = Math.min(
    CONSENSUS_QUORUM,
    Math.max(1, realPlayerCount),
  );

  // The submitter's standings must be the plurality AND meet quorum.
  const consensusReached =
    bestHash === standingsHash && matchingVotes >= effectiveQuorum;

  if (!consensusReached) {
    // Either the submitter disagrees with the plurality (likely forged), or we
    // don't yet have enough corroborating submissions.
    if (bestHash !== standingsHash && bestVotes >= effectiveQuorum) {
      // A different standings already reached quorum → this submission is a
      // minority (probably tampered). Reject outright.
      logger.warn(
        `Player ${playerId} standings disagree with quorum for ${tournamentId} (their votes=${matchingVotes}, quorum hash votes=${bestVotes}).`,
      );
      return { status: "rejected", reason: "Result did not match bracket consensus." };
    }
    // Not enough corroboration yet — keep the submission and ask the client to retry.
    return {
      status: "pending",
      have: matchingVotes,
      need: effectiveQuorum,
    };
  }

  if (realPlayerCount < CONSENSUS_QUORUM) {
    logger.info(
      `Bracket ${tournamentId} has ${realPlayerCount} real player(s); scoring ${playerId} on single-submission trust.`,
    );
  }

  // ── Consensus reached (or tiny-bracket fallback): server writes the score. ──
  const weeklyScore = BRACKET_SIZE + 1 - placement; // 1st → 32, 32nd → 1

  try {
    await writeLeaderboardScore(context, projectId, WEEKLY_LEADERBOARD_ID, playerId, weeklyScore, logger);
  } catch (e) {
    logger.error(`Weekly score write failed for ${playerId}: ${e.message}`);
    return { status: "rejected", reason: "Server error writing score." };
  }

  // ── Authoritative all-time championship counter (placement 1 only) ──
  if (placement === 1) {
    try {
      const champCount = await bumpChampionships(cloudSave, projectId, playerId, logger);
      await writeLeaderboardScore(
        context,
        projectId,
        ALLTIME_LEADERBOARD_ID,
        playerId,
        champCount,
        logger,
      );
    } catch (e) {
      logger.error(`All-time write failed for ${playerId}: ${e.message}`);
      // Weekly already written; do not fail the whole call for the all-time bump.
    }
  }

  // ── Mark scored so re-submits are no-ops ──
  try {
    await cloudSave.setPrivateCustomItem(projectId, scoredEntity, {
      key: playerId,
      value: "done",
    });
  } catch (e) {
    logger.error(`Failed to mark ${playerId} scored: ${e.message}`);
  }

  logger.info(
    `Scored ${playerId}: ${tournamentId} placement ${placement} → weekly ${weeklyScore}.`,
  );
  return { status: "scored", placement, score: weeklyScore };
};

// ═══════════════════════════════════════════════════════════════
//  Leaderboard write — VERIFY THIS AGAINST YOUR INSTALLED SDK VERSION.
//  The UGS Leaderboards Cloud Code SDK is versioned; the method name and
//  argument shape below match @unity-services/leaderboards-1.x. If your project
//  uses a different major version, adjust here (this is the ONLY place the
//  leaderboard write happens).
// ═══════════════════════════════════════════════════════════════

async function writeLeaderboardScore(
  context,
  projectId,
  leaderboardId,
  playerId,
  score,
  logger,
) {
  const { LeaderboardsApi } = require("@unity-services/leaderboards-1.4");
  const leaderboards = new LeaderboardsApi(context);
  await leaderboards.addLeaderboardPlayerScore(projectId, leaderboardId, playerId, {
    score,
  });
}

// ═══════════════════════════════════════════════════════════════
//  Cloud Save helpers
// ═══════════════════════════════════════════════════════════════

async function bumpChampionships(cloudSave, projectId, playerId, logger) {
  const entity = "arena_championships";
  let count = 0;
  const existing = await getItem(cloudSave, projectId, entity, playerId, logger);
  if (existing != null) {
    const n = parseInt(existing, 10);
    if (!isNaN(n) && n > 0) count = n;
  }
  count += 1;
  await cloudSave.setPrivateCustomItem(projectId, entity, {
    key: playerId,
    value: String(count),
  });
  return count;
}

async function getBracketRealPlayerCount(cloudSave, projectId, weekId, tournamentId, logger) {
  const bracketsEntity = `tournament_brackets_${weekId}`;
  const raw = await getItem(cloudSave, projectId, bracketsEntity, tournamentId, logger);
  if (raw == null) return 0;
  try {
    const b = JSON.parse(raw);
    if (typeof b.realPlayerCount === "number") return b.realPlayerCount;
    if (Array.isArray(b.realEntries)) return b.realEntries.length;
  } catch (e) {
    logger.error(`Bad bracket data for ${tournamentId}: ${e.message}`);
  }
  return 0;
}

async function hasItem(cloudSave, projectId, entity, key, logger) {
  return (await getItem(cloudSave, projectId, entity, key, logger)) != null;
}

async function getItem(cloudSave, projectId, entity, key, logger) {
  try {
    const res = await cloudSave.getPrivateCustomItems(projectId, entity);
    if (res?.data?.results) {
      const item = res.data.results.find((r) => r.key === key);
      if (item) return item.value;
    }
  } catch (e) {
    const status = e?.response?.status || e?.status;
    if (status !== 404) {
      logger.error(`getItem(${entity},${key}) error: ${e.message}`);
    }
  }
  return null;
}

async function getAllItems(cloudSave, projectId, entity, logger) {
  try {
    const res = await cloudSave.getPrivateCustomItems(projectId, entity);
    if (res?.data?.results) return res.data.results;
  } catch (e) {
    const status = e?.response?.status || e?.status;
    if (status !== 404) {
      logger.error(`getAllItems(${entity}) error: ${e.message}`);
    }
  }
  return [];
}

// Deterministic 32-bit string hash (Java hashCode equivalent) — matches the
// hashing style already used in LockAndAssignBrackets.js.
function hashString(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = (Math.imul(31, hash) + str.charCodeAt(i)) | 0;
  }
  return hash;
}

File: ArenaModule/ArenaModule/Handlers/SubmitTournamentResult.cs
// ArenaModule — SubmitTournamentResult handler
//
// Server-authoritative tournament scoring. Closes the "client writes its own
// leaderboard score" hole: the client no longer calls the Leaderboards API. It
// submits the FULL final standings it computed (ordered entryIds, champion
// first) plus its own claimed placement, and the server:
//
//   1. verifies the caller belongs to the tournament (via the server-authored
//      tournament_player_map_{weekId}) with the claimed entryId, and that the
//      claimed placement matches the player's position in the submitted standings;
//   2. cross-checks the submitted standings against OTHER real players in the
//      same bracket. Because every honest client resolves the identical
//      deterministic bracket, honest submissions share a standings hash; a lone
//      cheater's forged standings never reaches quorum and is rejected;
//   3. once a standings hash reaches quorum (or the bracket is too small for
//      quorum to exist), the server itself writes the weekly placement score and
//      maintains an authoritative all-time championship counter.
//
// Deploy with a dashboard Access-Control policy that denies player tokens write
// access to arena_weekly / arena_alltime, so this handler is the only writer.
//
// Cloud Save entities:
//   "tournament_player_map_{weekId}" — key playerId  → PlayerMapping     (written by LockAndAssignBrackets)
//   "tournament_brackets_{weekId}"   — key tournamentId → BracketData     (has RealPlayerCount)
//   "tournament_results_{weekId}"    — key playerId  → StoredResult
//   "tournament_scored_{weekId}"     — key playerId  → "done"             (idempotency guard)
//   "arena_championships"            — key playerId  → career championship count (authoritative)

using System.Text.Json;
using ArenaModule.Helpers;
using Microsoft.Extensions.Logging;
using Unity.Services.CloudCode.Apis;
using Unity.Services.CloudCode.Core;
using Unity.Services.CloudCode.Shared;
using Unity.Services.CloudSave.Model;
using Unity.Services.Leaderboards.Model;

namespace ArenaModule.Handlers;

public class SubmitTournamentResult
{
    private const string WeeklyLeaderboardId = "arena_weekly";
    private const string AllTimeLeaderboardId = "arena_alltime";
    private const string ChampionshipsEntity = "arena_championships";

    private const int BracketSize = 32;

    // Minimum real players (including the submitter) whose independently computed
    // standings must agree before a result is trusted. Brackets with fewer real
    // players than this can never reach quorum and fall back to single-submission
    // trust (logged).
    private const int ConsensusQuorum = 2;

    private readonly ILogger<SubmitTournamentResult> _logger;

    public SubmitTournamentResult(ILogger<SubmitTournamentResult> logger)
    {
        _logger = logger;
    }

    [CloudCodeFunction("SubmitTournamentResult")]
    public async Task<SubmitResultOutcome> Execute(
        IExecutionContext ctx,
        IGameApiClient api,
        string weekId,
        string tournamentId,
        string entryId,
        int placement,
        string standingsJson)
    {
        // ── Parameter validation ──
        if (string.IsNullOrWhiteSpace(weekId) || !weekId.StartsWith('W'))
            return Rejected("Invalid week ID.");
        if (string.IsNullOrWhiteSpace(tournamentId))
            return Rejected("Invalid tournament ID.");
        if (string.IsNullOrWhiteSpace(entryId))
            return Rejected("Invalid entry ID.");
        if (placement < 1 || placement > BracketSize)
            return Rejected("Invalid placement.");
        if (string.IsNullOrWhiteSpace(standingsJson))
            return Rejected("Standings are required.");

        List<string>? standings;
        try
        {
            standings = JsonSerializer.Deserialize<List<string>>(standingsJson);
        }
        catch
        {
            return Rejected("Malformed standings.");
        }
        if (standings == null || standings.Count == 0)
            return Rejected("Standings must be a non-empty array.");

        var playerId = ctx.PlayerId;

        // ── Idempotency: already scored this week? ──
        var scoredEntity = $"tournament_scored_{weekId}";
        if (await GetItem(api, ctx, scoredEntity, playerId) != null)
            return new SubmitResultOutcome { Status = "already_scored" };

        // ── Authority: the player must be mapped to THIS tournament with THIS entryId ──
        var mapEntity = $"tournament_player_map_{weekId}";
        var mappingRaw = await GetItem(api, ctx, mapEntity, playerId);
        if (mappingRaw == null)
            return Rejected("You are not registered for this week.");

        PlayerMapping? mapping;
        try { mapping = JsonSerializer.Deserialize<PlayerMapping>(mappingRaw); }
        catch { return Rejected("Server mapping unreadable."); }

        if (mapping == null || mapping.TournamentId != tournamentId || mapping.EntryId != entryId)
            return Rejected("Tournament/entry mismatch.");

        // ── Internal consistency: claimed placement must match the player's own
        //    position in the submitted standings. ──
        int claimedIndex = standings.IndexOf(entryId);
        if (claimedIndex == -1)
            return Rejected("Your entry is absent from the standings.");
        if (claimedIndex + 1 != placement)
            return Rejected("Claimed placement disagrees with submitted standings.");

        // ── Store this submission (normalized hash so formatting can't split votes) ──
        int standingsHash = ScheduleHelper.HashCode(JsonSerializer.Serialize(standings));
        var resultsEntity = $"tournament_results_{weekId}";
        var stored = new StoredResult
        {
            TournamentId = tournamentId,
            StandingsHash = standingsHash,
            Placement = placement,
            EntryId = entryId,
            SubmittedUtc = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        };
        try
        {
            await api.CloudSaveData.SetCustomItemAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, resultsEntity,
                new SetItemBody(playerId, JsonSerializer.Serialize(stored)));
        }
        catch (Exception ex)
        {
            _logger.LogError("Failed to store result for {PlayerId}: {Error}", playerId, ex.Message);
            return Rejected("Server error saving result.");
        }

        // ── Tally standings hashes across all submissions for this tournament ──
        int matchingVotes = 0;
        int bestVotes = 0;
        int bestHash = 0;
        var hashVotes = new Dictionary<int, int>();

        foreach (var value in await GetAllValues(api, ctx, resultsEntity))
        {
            StoredResult? r;
            try { r = JsonSerializer.Deserialize<StoredResult>(value); }
            catch { continue; }
            if (r == null || r.TournamentId != tournamentId) continue;

            hashVotes.TryGetValue(r.StandingsHash, out int v);
            hashVotes[r.StandingsHash] = v + 1;
            if (r.StandingsHash == standingsHash) matchingVotes++;
        }
        foreach (var kv in hashVotes)
        {
            if (kv.Value > bestVotes) { bestVotes = kv.Value; bestHash = kv.Key; }
        }

        int realPlayerCount = await GetBracketRealPlayerCount(api, ctx, weekId, tournamentId);
        int effectiveQuorum = Math.Min(ConsensusQuorum, Math.Max(1, realPlayerCount));

        bool consensusReached = bestHash == standingsHash && matchingVotes >= effectiveQuorum;

        if (!consensusReached)
        {
            if (bestHash != standingsHash && bestVotes >= effectiveQuorum)
            {
                // A different standings already reached quorum → this is a minority
                // (probably tampered) submission. Reject.
                _logger.LogWarning(
                    "Player {PlayerId} standings disagree with quorum for {Tournament} (theirs={Mine}, quorum={Quorum}).",
                    playerId, tournamentId, matchingVotes, bestVotes);
                return Rejected("Result did not match bracket consensus.");
            }
            // Not enough corroboration yet — keep the submission and ask the client to retry.
            return new SubmitResultOutcome { Status = "pending", Placement = placement };
        }

        if (realPlayerCount < ConsensusQuorum)
        {
            _logger.LogInformation(
                "Bracket {Tournament} has {Count} real player(s); scoring {PlayerId} on single-submission trust.",
                tournamentId, realPlayerCount, playerId);
        }

        // ── Consensus reached: server writes the score. ──
        int weeklyScore = BracketSize + 1 - placement; // 1st → 32, 32nd → 1

        try
        {
            await WriteLeaderboardScore(api, ctx, WeeklyLeaderboardId, playerId, weeklyScore);
        }
        catch (Exception ex)
        {
            _logger.LogError("Weekly score write failed for {PlayerId}: {Error}", playerId, ex.Message);
            return Rejected("Server error writing score.");
        }

        if (placement == 1)
        {
            try
            {
                int champCount = await BumpChampionships(api, ctx, playerId);
                await WriteLeaderboardScore(api, ctx, AllTimeLeaderboardId, playerId, champCount);
            }
            catch (Exception ex)
            {
                // Weekly already written; don't fail the whole call for the all-time bump.
                _logger.LogError("All-time write failed for {PlayerId}: {Error}", playerId, ex.Message);
            }
        }

        // ── Mark scored so re-submits are no-ops ──
        try
        {
            await api.CloudSaveData.SetCustomItemAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, scoredEntity,
                new SetItemBody(playerId, "done"));
        }
        catch (Exception ex)
        {
            _logger.LogError("Failed to mark {PlayerId} scored: {Error}", playerId, ex.Message);
        }

        _logger.LogInformation(
            "Scored {PlayerId}: {Tournament} placement {Placement} → weekly {Score}.",
            playerId, tournamentId, placement, weeklyScore);

        return new SubmitResultOutcome { Status = "scored", Placement = placement, Score = weeklyScore };
    }

    // ═══════════════════════════════════════════════════════════
    //  Leaderboard write (server-authoritative)
    // ═══════════════════════════════════════════════════════════

    private static async Task WriteLeaderboardScore(
        IGameApiClient api, IExecutionContext ctx, string leaderboardId, string playerId, double score)
    {
        await api.Leaderboards.AddLeaderboardPlayerScoreAsync(
            ctx,
            ctx.ServiceToken,
            Guid.Parse(ctx.ProjectId),
            leaderboardId,
            playerId,
            new LeaderboardScore(score));
    }

    // ═══════════════════════════════════════════════════════════
    //  Cloud Save helpers
    // ═══════════════════════════════════════════════════════════

    private async Task<int> BumpChampionships(IGameApiClient api, IExecutionContext ctx, string playerId)
    {
        int count = 0;
        var existing = await GetItem(api, ctx, ChampionshipsEntity, playerId);
        if (existing != null && int.TryParse(existing, out int n) && n > 0)
            count = n;
        count += 1;
        await api.CloudSaveData.SetCustomItemAsync(
            ctx, ctx.ServiceToken, ctx.ProjectId, ChampionshipsEntity,
            new SetItemBody(playerId, count.ToString()));
        return count;
    }

    private async Task<int> GetBracketRealPlayerCount(
        IGameApiClient api, IExecutionContext ctx, string weekId, string tournamentId)
    {
        var raw = await GetItem(api, ctx, $"tournament_brackets_{weekId}", tournamentId);
        if (raw == null) return 0;
        try
        {
            var b = JsonSerializer.Deserialize<BracketData>(raw);
            if (b == null) return 0;
            if (b.RealPlayerCount > 0) return b.RealPlayerCount;
            return b.RealEntries?.Count ?? 0;
        }
        catch (Exception ex)
        {
            _logger.LogError("Bad bracket data for {Tournament}: {Error}", tournamentId, ex.Message);
            return 0;
        }
    }

    /// <summary>Returns the value for a single key in an entity, or null.</summary>
    private async Task<string?> GetItem(IGameApiClient api, IExecutionContext ctx, string entity, string key)
    {
        try
        {
            var res = await api.CloudSaveData.GetCustomItemsAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, entity);
            if (res?.Data?.Results != null)
            {
                foreach (var item in res.Data.Results)
                {
                    if (item.Key == key)
                        return item.Value?.ToString();
                }
            }
        }
        catch (ApiException ex) when (ex.Response?.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            // entity doesn't exist yet
        }
        catch (Exception ex)
        {
            _logger.LogError("GetItem({Entity},{Key}) error: {Error}", entity, key, ex.Message);
        }
        return null;
    }

    /// <summary>Returns all item values in an entity (JSON strings).</summary>
    private async Task<List<string>> GetAllValues(IGameApiClient api, IExecutionContext ctx, string entity)
    {
        var values = new List<string>();
        try
        {
            var res = await api.CloudSaveData.GetCustomItemsAsync(
                ctx, ctx.ServiceToken, ctx.ProjectId, entity);
            if (res?.Data?.Results != null)
            {
                foreach (var item in res.Data.Results)
                {
                    var v = item.Value?.ToString();
                    if (!string.IsNullOrEmpty(v)) values.Add(v);
                }
            }
        }
        catch (ApiException ex) when (ex.Response?.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            // entity doesn't exist yet
        }
        catch (Exception ex)
        {
            _logger.LogError("GetAllValues({Entity}) error: {Error}", entity, ex.Message);
        }
        return values;
    }

    private static SubmitResultOutcome Rejected(string reason)
        => new() { Status = "rejected", Reason = reason };
}
